kinesis/src/main/scala/com.snowplowanalytics.snowplow.collectors.scalastream/sinks/KinesisSink.scala

```diff
@@ -42,9 +39,322 @@ import cats.syntax.either._
+            }
+          case Failure(f) =>
+            log.error("Writing failed with error:\n", f)
+            val errorMessage = {
+              if (buffer.isEmpty || retriesLeft > 1)
+                s"Retrying writing batch to Kinesis stream $streamName in $nextBackoff milliseconds..."
+              else
+                s"Sending batch to SQS queue ${buffer.get.sqsBufferName} in $nextBackoff milliseconds..."
+            }
+            log.error(errorMessage)
+            scheduleWrite(batch, target, nextBackoff)(retriesLeft - 1)
+        }
+
+      case SqsQueue(sqs) =>
+        log.info(s"Writing ${batch.size} Thrift records to SQS queue ${sqs.sqsBufferName}.")
+        writeBatchToSqs(batch, sqs).onComplete {
+          case Success(s) =>
+            log.info(
+              s"Successfully wrote ${batch.size - s.size} out of ${batch.size} messages to SQS queue ${sqs.sqsBufferName}."
+            )
+            if (s.nonEmpty) {
+              s.foreach { f =>
+                log.error(
+                  s"Failed writing message to SQS queue ${sqs.sqsBufferName}, with error code [${f._2.code}] and message [${f._2.message}]."
+                )
+              }
+              val errorMessage = {
+                if (retriesLeft > 1)
+                  s"Retrying all messages that could not be written to SQS queue ${sqs.sqsBufferName} in $nextBackoff milliseconds..."
+                else
+                  s"Sending ${s.size} messages that could not be written to SQS queue ${sqs.sqsBufferName} to Kinesis stream $streamName in $nextBackoff milliseconds..."
+              }
+              log.error(errorMessage)
+              scheduleWrite(s.map(_._1), target, nextBackoff)(retriesLeft - 1)
+            }
+          case Failure(f) =>
+            log.error("Writing failed with error:\n", f)
+            val errorMessage = {
+              if (retriesLeft > 1)
+                s"Retrying writing batch to SQS queue ${sqs.sqsBufferName} in $nextBackoff milliseconds..."
+              else
+                s"Sending batch to Kinesis stream $streamName in $nextBackoff milliseconds..."
+            }
+            log.error(errorMessage)
+            scheduleWrite(batch, target, nextBackoff)(retriesLeft - 1)
+        }
+    }
+
+  def writeBatchToKinesis(batch: List[Events]): Future[PutRecordsResult] =
+    Future {
+      val putRecordsRequest = {
+        val prr = new PutRecordsRequest()
+        prr.setStreamName(streamName)
+        val putRecordsRequestEntryList = batch.map { event =>
+          val prre = new PutRecordsRequestEntry()
+          prre.setPartitionKey(event.key)
+          prre.setData(ByteBuffer.wrap(event.payloads))
+          prre
+        }
+        prr.setRecords(putRecordsRequestEntryList.asJava)
+        prr
+      }
+      client.putRecords(putRecordsRequest)
+    }
+
+  def writeBatchToSqs(batch: List[Events], sqs: SqsClientAndName): Future[List[(Events, BatchResultErrorInfo)]] =
```

Comment:
Would be nice to add a comment to say that this function returns failed inserts.